### PR TITLE
[Parse] Tokens.def cleanup and documentation

### PR DESCRIPTION
**Pull request by micahhainline**
This request cleans up the internal structure of Tokens.def, introduces some minor abstraction layers for ease of use and consistency, and documents usage. It also updates usage for consistency. 

This was discussed briefly but with only positive feedback on the [swift-dev](https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20161226/003756.html) mailing list.

### CODE REVIEW

**Comment by micahhainline** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
LGTM.  Thanks!

@swift-ci please smoke test.

File: include/swift/Parse/Tokens.def [hunk 1]
```diff
@@ -9,59 +9,100 @@
...
 //===----------------------------------------------------------------------===//
-//
-// This file defines macros used for macro-metaprogramming lexer tokens.
-//
+///
+/// This file defines macros used for macro-metaprogramming lexer tokens.
+///
+/// To use, simply define one or more of the macro functions listed below and
+/// then import this file. The macro functions follow a hierarchical pattern,
+/// and defining a higher-level macro will expend (by default) all the patterns
```
Comment (xwu) [inline include/swift/Parse/Tokens.def]: s/expend/expand/

File: include/swift/Parse/Tokens.def [hunk 2]
```diff
@@ -9,59 +9,100 @@
...
 /// POUND_OBJECT_LITERAL(kw, desc, proto)
-///   Every keyword in the #foo namespace representing an object literal.
+///   Every keyword in the # namespace representing a new-form object literal.
 #ifndef POUND_OBJECT_LITERAL
 #define POUND_OBJECT_LITERAL(kw, desc, proto) POUND_KEYWORD(kw)
 #endif
 
 /// POUND_OLD_OBJECT_LITERAL(kw, new_kw, old_arg, new_arg)
-///   Every keyword in the #foo namespace representing an object literal.
+///   Every keyword in the # namespace representing an old-form object literal.
```
Comment (xwu) [inline include/swift/Parse/Tokens.def]: This is a good addition, but perhaps it'd be better to explicitly say it's obsolete; it's not one of two valid ways of spelling an object literal, it's just here for migrator purposes.

File: include/swift/Parse/Tokens.def [hunk 3]
```diff
@@ -9,59 +9,100 @@
...
 /// POUND_KEYWORD(kw)
-///   Every keyword in the #foo namespace.
+///   Every keyword in the # namespace.
 #ifndef POUND_KEYWORD
 #define POUND_KEYWORD(kw)
 #endif
 
 /// POUND_OBJECT_LITERAL(kw, desc, proto)
-///   Every keyword in the #foo namespace representing an object literal.
+///   Every keyword in the # namespace representing a new-form object literal.
```
Comment (xwu) [inline include/swift/Parse/Tokens.def]: Following up on the comment below, I'd say that this is the syntax for an object literal. "New-form" is not a term used elsewhere to refer to the current syntax, and it implies that there is another form in the language, which there isn't.

File: include/swift/Parse/Tokens.def [hunk 4]
```diff
@@ -9,59 +9,100 @@
...
+///   'false', and 'as'
+#ifndef EXPR_KEYWORD
+#define EXPR_KEYWORD(kw) SWIFT_KEYWORD(kw)
+#endif
+
+/// PAT_KEYWORD(kw)
+///   Expands for every Swift keyword used in a pattern, which is currently
+///   limited to '_'
+#ifndef PAT_KEYWORD
+#define PAT_KEYWORD(kw) SWIFT_KEYWORD(kw)
```
Comment (xwu) [inline include/swift/Parse/Tokens.def]: Question about the taxonomy here: isn't every pattern keyword an expression keyword, and every expression keyword a statement keyword?

**Comment by slavapestov** [comment]
@DougGregor or @jckarter can you take a look?

**Comment by CodaFi** [comment]
These changes were merged into master on #10034

Thank you for your contribution @micahhainline.  I'm apologize for not getting this reviewed sooner.

**Pull request closed without merging**